CC-16934 Glue support for Unzer IG. (#1167)
spryker/spryker-docs: docs/pbc/all/payment-service-providers/unzer/install-unzer/integrate-unzer-glue.md

{
    /**
     * @return array<\Spryker\Glue\CheckoutRestApiExtension\Dependency\Plugin\CheckoutDataResponseMapperPluginInterface>
     */
    protected function getCheckoutDataResponseMapperPlugins(): array
    {
        return [
            new UnzerCheckoutDataResponseMapperPlugin(),
        ];
    }
}
```

**src/Pyz/Glue/GlueStorefrontApiApplication/GlueStorefrontApiApplicationDependencyProvider.php**

```php
namespace Pyz\Glue\GlueStorefrontApiApplication;

use Spryker\Glue\GlueStorefrontApiApplication\GlueStorefrontApiApplicationDependencyProvider as SprykerGlueStorefrontApiApplicationDependencyProvider;
use SprykerEco\Glue\UnzerRestApi\Plugin\GlueJsonApiConventionExtension\UnzerNotificationResource;

class GlueStorefrontApiApplicationDependencyProvider extends SprykerGlueStorefrontApiApplicationDependencyProvider
{
    /**
     * @return array<\Spryker\Glue\GlueApplicationExtension\Dependency\Plugin\ResourceInterface>
     */
    protected function getResourcePlugins(): array
    {
        return [
            new UnzerNotificationResource(),
        ];
    }
}
```

{% info_block warningBox "Verification" %}

> 

Reviewer: Add an introductory sentence explaining what we verify—for example, "Verify that the following API requests work:"

1. Create a cart by sending the `POST https://glue.mysprykershop.com/carts`.
2. Add at least one item to the cart by sending `POST https://glue.mysprykershop.com/{{cart_uuid}}/items`.
3. Check result by sending the `POST https://glue.mysprykershop.com/checkout-data?include=payment-methods` request.

**Request Example:**
```json
{
  "data":
  {
    "type": "checkout-data",
    "attributes":
    {
      "idCart": "{{cart_uuid}}",
      "payments": [
        {
          "paymentMethodName": "Unzer Sofort",
          "paymentProviderName": "Unzer"
        }
      ]
    }
  }
}
```

> {% info_block infoBox %}

Reviewer: If the steps listed in this infoBox are mandatory, then replace `{% info_block infoBox %}` with  `{% info_block warningBox %}`

> Take care that requests body differs for each Unzer payment method.

Reviewer: ```suggestion
Ensure that the request body differs for each Unzer payment method.
```

> - Property `paymentMethodName` of `payments` has to be replaced by the used method—for example, `Unzer Sofort` or `Unzer Credit Card`. 

Reviewer: ```suggestion
- Property `paymentMethodName` of `payments` must be replaced by the used method—for example, `Unzer Sofort` or `Unzer Credit Card`. 
```

- Property `paymentResource` of `payments` is not required for `Unzer Sofort`, `Unzer Marketplace Sofort`, and `Unzer Bank Transfer`.

{% endinfo_block %}

> **Response Example:**

Reviewer: ```suggestion
<details><summary markdown='span'>Response example:</summary>
```

```json
{
  "data": {
    "type": "checkout-data",
    "id": null,
    "attributes": {
      "unzerPublicKey": "string",
      ...
    },
    "links": {
      "self": "https://glue.mysprykershop.com/checkout-data?include=payment-methods"
    },
    "relationships": {
      "payment-methods": {
        "data": [
          {
            "type": "payment-methods",
            "id": "1"
          }
        ]
      }
    },
    ...
  },
  "included": [
    {
      "type": "payment-methods",
      "id": "1",
      "attributes": {
        "paymentMethodName": "Unzer Sofort",
        "paymentProviderName": "Unzer",
        "priority": 1,
        "requiredRequestData": [
          "paymentMethod",
          "paymentProvider",
          "unzerPayment.paymentResource.id"
        ]
      },
      "links": {
        "self": "https://glue.mysprykershop.com/payment-methods/1"
      }
    },
    ...
  ]
}
```

> 

Reviewer: ```suggestion

</details>

```

4. Check result by sending the `POST https://glue.mysprykershop.com/checkout` request.

**Request Example:**
```json
{
  "data": {
    "type": "checkout",
    "attributes": {
      "idCart": "{{cart_uuid}}",
      "payments": [
        {
          "paymentMethodName": "Unzer Credit Card",
          "paymentProviderName": "Unzer",
          "paymentSelection": "unzerCreditCard",
          "unzerCreditCard": {
            "paymentResource": {
              "id": "{{payment_resource}}"
            }
          },
          "amount": "{{amount}}"
        }
      ],
      ...
    }
  }
}
```

> {% info_block infoBox %}

Reviewer: If the steps listed in this infoBox are mandatory, then replace `{% info_block infoBox %}` with  `{% info_block warningBox %}`

> Take care that requests body differs for each Unzer payment method.

Reviewer: ```suggestion
Ensure that the request body differs for each Unzer payment method:
```

> - Property `paymentMethodName` of `payments` has to be replaced by used method (e.g. `Unzer Sofort`, `Unzer Credit Card`, etc.).

Reviewer: ```suggestion
- Property `paymentMethodName` of `payments` must be replaced by the used method—for example, `Unzer Sofort` or `Unzer Credit Card`.
```

> - Property `paymentResource` of `payments` is not required for `Unzer Sofort`, `Unzer Marketplace Sofort`, `Unzer Bank Transfer`.

Reviewer: ```suggestion
- Property `paymentResource` of `payments` is not required for `Unzer Sofort`, `Unzer Marketplace Sofort`, and  `Unzer Bank Transfer`.
```

> - Property `paymentResource` of `payments` is not required for `Unzer Sofort`, `Unzer Marketplace Sofort`, `Unzer Bank Transfer`.

Reviewer: ```suggestion
- Property `paymentResource` of `payments` is not required for `Unzer Sofort`, `Unzer Marketplace Sofort`, and  `Unzer Bank Transfer`.
```

> - Property `paymentSelection` of `payments` has to be replaced by used method (e.g. `unzerSofort`, `unzerCreditCard`, `unzerMarketplaceCreditCard`, etc.).

Reviewer: ```suggestion
- Property `paymentSelection` of `payments` has to be replaced by used method—for example, `unzerSofort`, `unzerCreditCard`, or `unzerMarketplaceCreditCard`.
```

| Unzer               | [Unzer feature integration](/docs/pbc/all/payment-service-providers/unzer/install-unzer/integrate-unzer.html)                                            |
| Glue API - Checkout | [Glue API - Checkout feature integration](/docs/scos/dev/feature-integration-guides/202204.0/glue-api/glue-api-checkout-feature-integration.html) |
| Glue API - Payments | [Glue API - Payments feature integration](/docs/scos/dev/feature-integration-guides/202204.0/glue-api/glue-api-payments-feature-integration.html) |

### 1) Install the required modules using Composer

Install the required modules:

```bash
composer require spryker-eco/unzer-rest-api
```

{% info_block warningBox "Verification" %}

Make sure that the following modules have been installed:

| MODULE       | EXPECTED DIRECTORY                 |
|--------------|------------------------------------|
| UnzerRestApi | vendor/spryker-eco/unzer-rest-api  |

{% endinfo_block %}

---

### 2) Set up the configuration

Put all the payment methods available in the shop to  `PaymentsRestApiConfig`:

> **src/Pyz/Glue/CheckoutRestApi/CheckoutRestApiConfig.php**

Reviewer 2: file looks wrong

        ];
    }
}
```

**src/Pyz/Glue/GlueStorefrontApiApplication/GlueStorefrontApiApplicationDependencyProvider.php**

```php
namespace Pyz\Glue\GlueStorefrontApiApplication;

use Spryker\Glue\GlueStorefrontApiApplication\GlueStorefrontApiApplicationDependencyProvider as SprykerGlueStorefrontApiApplicationDependencyProvider;
use SprykerEco\Glue\UnzerRestApi\Plugin\GlueJsonApiConventionExtension\UnzerNotificationResource;

class GlueStorefrontApiApplicationDependencyProvider extends SprykerGlueStorefrontApiApplicationDependencyProvider
{
    /**
     * @return array<\Spryker\Glue\GlueApplicationExtension\Dependency\Plugin\ResourceInterface>
     */
    protected function getResourcePlugins(): array
    {
        return [
            new UnzerNotificationResource(),
        ];
    }
}
```

{% info_block warningBox "Verification" %}

1. Create a cart by sending the `POST https://glue.mysprykershop.com/carts`.
2. Add at least one item to the cart by sending `POST https://glue.mysprykershop.com/{{cart_uuid}}/items`.
3. Check result by sending the `POST https://glue.mysprykershop.com/checkout-data?include=payment-methods` request.

> **Request:**

Reviewer 2: ```suggestion
**Request Example:**
```

```json
{
  "data":
  {
    "type": "checkout-data",
    "attributes":
    {
      "idCart": "{{cart_uuid}}",
      "payments": [
        {
          "paymentMethodName": "Unzer Sofort",
          "paymentProviderName": "Unzer"
        }
      ]
    }
  }
}
```

{% info_block infoBox %}

Please take care that requests body differs for each Unzer payment method.
- Property `paymentMethodName` of `payments` has to be replaced by used method (e.g. `Unzer Sofort`, `Unzer Credit Card`, etc.). 
- Property `paymentResource` of `payments` is not required for `Unzer Sofort`, `Unzer Marketplace Sofort`, `Unzer Bank Transfer`.

{% endinfo_block %}

> **Response:**

Reviewer 2: ```suggestion
**Response Example:**
```

```json
{
  "data": {
    "type": "checkout-data",
    "id": null,
    "attributes": {
      "unzerPublicKey": "string",
      ...
    },
    "links": {
      "self": "https://glue.mysprykershop.com/checkout-data?include=payment-methods"
    },
    "relationships": {
      "payment-methods": {
        "data": [
          {
            "type": "payment-methods",
            "id": "1"
          }
        ]
      }
    },
    ...
  },
  "included": [
    {
      "type": "payment-methods",
      "id": "1",
      "attributes": {
        "paymentMethodName": "Unzer Sofort",
        "paymentProviderName": "Unzer",
        "priority": 1,
        "requiredRequestData": [
          "paymentMethod",
          "paymentProvider",
          "unzerPayment.paymentResource.id"
        ]
      },
      "links": {
        "self": "https://glue.mysprykershop.com/payment-methods/1"
      }
    },
    ...
  ]
}
```

4. Check result by sending the `POST https://glue.mysprykershop.com/checkout` request.

> **Request:**

Reviewer 2: same here and below

---
title: Glue support in Unzer feature integration
description: Integrate the Glue support in Unzer into your project
last_updated: Aug 10, 2022
template: feature-integration-guide-template
---

This document describes how to integrate the *Glue support* to the [Unzer](/docs/pbc/all/payment-service-providers/unzer/unzer.html) into a Spryker project.

## Install feature core

> Follow the steps below to install the Glue for Unzer.

Reviewer: I think this sentence is even better.
```suggestion
Follow the steps below to install the Unzer Glue API feature core.
```
